pio-tools/custom_target.py

```diff
@@ -21,14 +21,15 @@
 import shutil
 import json
 from colorama import Fore, Back, Style
+from platformio.project.config import ProjectConfig
 
 Import("env")
 platform = env.PioPlatform()
 board = env.BoardConfig()
 mcu = board.get("build.mcu", "esp32")
+esptoolpy = os.path.join(ProjectConfig.get_instance().get("platformio", "packages_dir"),("tool-esptoolpy") or "", "esptool.py")
```

Comment:
pathlib is the more modern way since Python 3.4.

Reply:
Not here, since we need explicit the pio venv python version. So thats why using pio API